hadoop-hdds/docs/content/design/event-notifications.md

```diff
@@ -0,0 +1,193 @@
+
+## Open questions
+
+### Notification Schema
+
+The S3 event notification schema https://docs.aws.amazon.com/AmazonS3/latest/userguide/notification-how-to-event-types-and-destinations.html#supported-notification-event-types is based around immutable objects which cannot be renamed.
+In (AWS) S3 semantics renames are notified at the per-object level by sending a delete event on the old path and a create event on the new path.
+However, in Ozone objects can be renamed via the "RenameKey" ratis event which is triggered for both file and directory renames.
+In the context of a FSO directory rename we see a RenameKey ratis event for the directory name and not the granular objects.
+This would allow us to emit an event to denote that a path or directory was renamed but:
+
+It does not distinguish between file vs directory renames
+It does not allow us to emit cascading delete / create events for all the child objects being deleted/created as a result of a directory rename operation (as would be expected by traditional S3 event semantics)
+
+This leaves us with a gap with two potential solutions:
+
+1. Add a bespoke/“extended” event S3 type for RENAME operations which simply conveys that the key has been renamed (as per the semantics of RenameKey ratis callback) and which then leaves it up to the event consumer to deem the context (file vs directory) themselves (e.g. by stat-ing the target path).  This is in-line with the semantics of HDFS inotify rename events:
+
+https://hadoop.apache.org/docs/stable/api/org/apache/hadoop/hdfs/inotify/Event.RenameEvent.html
+
+Pros:
+* does not require any new processing on the OM side
+
+Cons:
+* non standard S3 event semantics
+
+2. The Ozone manager has to somehow pull together the child objects impacted by directory RENAME and emit appropriate remove/create events for the source/destination paths.
+
+Pros:
+* standard S3 event semantics
+
+Cons:
+* required additional processing on the OM side to gather the necessary child objects impacted and emit events for RENAME operations.
+
+As with RENAMEs there is no concept for ACL events in S3.  If we stay with S3 semantics we could extend the schema similar to how Ceph has here: https://docs.ceph.com/en/latest/radosgw/s3-notification-compatibility/#event-types
+Or we if we took a HDFS Inotify view of the schema that already has inbuilt support for ACLs
+
+### At-least-once delivery
+
+At-least-once delivery cannot be guaranteed without some form of persistence, however we want to avoid persisting the notification messages themselves. One approach to achieving this is by "replaying" the proposals.
+If we persist the lastAppliedTxn after each notification is sent successfully, on a restart we could reset the lastAppliedTxn on the state machine and replay proposals in a "non-committal" mode in order to generate any missing notifications up to the current proposal.
```

Comment:
we do persist lastAppliedIndex in ratis log, and there's a callback to the statemachine implementation when a transaction is applied, so it sounds doable.

when we conceived the original design in HDDS-5984, we couldn't do it as S3 gateway is stateless. Making the logic inside the OM follower makes a lot of sense.